pennylane/visualize/visualize.py

```diff
@@ -0,0 +1,351 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from matplotlib import pyplot as plt
+from IPython import display
+
+
+def _j_nb():
+    try:
+        from IPython import get_ipython
+
+        if "IPKernelApp" not in get_ipython().config:
+            return False
+    except ImportError:
+        return False
+    except AttributeError:
+        return False
+    return True
+
+
+class Visualize:
+
+    """
+    A context manager for storing data relating to visualizations and post-processing of
+    variational quantum circuit optimization.
+
+    When a PennyLane optimization loop is wrapped in the Visualize context manager, data from each optimization step
+    is recorded and outputted as text or a graph.
+
+    Args:
+        steps (int): The number of steps for which the circuit is being optimized.
```

Comment:
I wonder why this is needed? The UI is still to run a separate loop for optimization. I tried locally to run the code 

```python
with Visualize(10, cost_fn=cost) as viz:
    for i in range(50):
        params = opt.step(cost, params)
```
and it works fine, except the print statements are wrong, e.g., 50 / 10